cpu/samd21/periph/adc.c

```diff
@@ -0,0 +1,361 @@
+    {
+        ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_NEG_INPUT);
+        ADC_0_PORT.PINCFG[ADC_0_NEG_INPUT].bit.INEN = true;
+        ADC_0_PORT.PINCFG[ADC_0_NEG_INPUT].bit.PULLEN = false;
+    } 
+
+    /* Set RUN_IN_STANDBY */
+    adc->CTRLA.reg = (ADC_0_RUN_IN_STANDBY << ADC_CTRLA_RUNSTDBY_Pos);
+
+    /* Set Voltage Reference */
+    adc->REFCTRL.reg = (ADC_0_REF_COM_EN << ADC_REFCTRL_REFCOMP_Pos) | ADC_0_REF_DEFAULT;
+
+    switch (precision)
+    {
+    case ADC_0_RES_8BIT:
+        resolution = ADC_0_RES_8BIT;
+        break;
+    case ADC_0_RES_10BIT:
+        resolution = ADC_0_RES_10BIT;
+        break;
+    case ADC_0_RES_12BIT:
+        resolution = ADC_0_RES_12BIT;
+        break;
+    case ADC_0_RES_16BIT:
+        divideResult = ADC_0_DIV_RES_DISABLE;
+        switch(ADC_0_ACCUM_DEFAULT)
+        {
+            case ADC_0_ACCUM_512:
+                accumulate = ADC_0_ACCUM_512;
+                break;
+            case ADC_0_ACCUM_1024:
+                accumulate = ADC_0_ACCUM_1024;
+                break;
+            default:
+                accumulate = ADC_0_ACCUM_256;
+                break;
+        }        
+        resolution = ADC_0_RES_16BIT;
+        break;
+    default:
+        DEBUG("Invalid precision");        
+        return -1;
+    }
+
+    /* Set the accumlation and devide result */
+    adc->AVGCTRL.reg = ADC_AVGCTRL_ADJRES(divideResult) | accumulate;
+
+    /* Set Sample length */
+    adc->SAMPCTRL.reg = (ADC_0_SAMPLE_LENGTH << ADC_SAMPCTRL_SAMPLEN_Pos);
+    while(adc_syncing(adc));
+    
+    /* If external vref. Pin setup */
+    if(ADC_0_REF_DEFAULT == ADC_0_REF_EXT_B)
+    {
+        ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_REF_DEFAULT);
+        ADC_0_PORT.PINCFG[ADC_0_REF_DEFAULT].bit.INEN = true;
+        ADC_0_PORT.PINCFG[ADC_0_REF_DEFAULT].bit.PULLEN = false;
+    }
+
+    /* Configure CTRLB Register HERE IS THE RESOLUTION SET!*/
+    adc->CTRLB.reg = 
+        ADC_0_PRESCALER |
+        resolution | 
+        (ADC_0_CORRECTION_EN << ADC_CTRLB_CORREN_Pos) | 
+        (ADC_0_FREE_RUNNING << ADC_CTRLB_FREERUN_Pos) | 
+        (ADC_0_LEFT_ADJUST << ADC_CTRLB_LEFTADJ_Pos) |
+        (ADC_0_DIFFERENTIAL_MODE << ADC_CTRLB_DIFFMODE_Pos);        
+
+
+    /* Configure Window Mode Register */
+    /*adc->WINCTRL.reg = ADC_0_WINDOW_MODE;
+    while(adc_syncing(adc));*/
+
+    /* Configure lower threshold */
+    adc->WINLT.reg = ADC_0_WINDOW_LOWER << ADC_WINLT_WINLT_Pos;
+    while(adc_syncing(adc));
+
+    /* Configure lower threshold */
+    adc->WINUT.reg = ADC_0_WINDOW_HIGHER << ADC_WINUT_WINUT_Pos;
+    while(adc_syncing(adc));
+
+    /* Configure PIN SCAN MODE & positive & Negative Input Pins */
+    adc->INPUTCTRL.reg = 
+        ADC_0_GAIN_FACTOR_DEFAULT |
+        (ADC_0_PIN_SCAN_OFFSET_START << ADC_INPUTCTRL_INPUTOFFSET_Pos) |
+        (ADC_0_PIN_SCAN_INPUT_TO_SCAN << ADC_INPUTCTRL_INPUTSCAN_Pos) |
+        ADC_0_NEG_INPUT |
+        ADC_0_POS_INPUT;
```

Comment:
Looking at the datasheet (table 30-14 p.827) `0x6` seems to be reserved. I don't really get why